lib/web/terminal.go

```diff
@@ -864,7 +821,141 @@ func serverHostPort(servername string) (string, int, error) {
 	return host, port, nil
 }
 
-func (t *TerminalHandler) write(data []byte, ws *websocket.Conn) (n int, err error) {
+// WithTerminalStreamEncoder overrides the default stream encoder
+func WithTerminalStreamEncoder(enc *encoding.Encoder) func(stream *TerminalStream) {
+	return func(stream *TerminalStream) {
+		stream.encoder = enc
+	}
+}
+
+// WithTerminalStreamDecoder overrides the default stream decoder
+func WithTerminalStreamDecoder(dec *encoding.Decoder) func(stream *TerminalStream) {
+	return func(stream *TerminalStream) {
+		stream.decoder = dec
+	}
+}
+
+// WithTerminalStreamResizeHandler provides a channel to subscribe to
+// terminal resize events
+func WithTerminalStreamResizeHandler(resizeC chan<- *session.TerminalParams) func(stream *TerminalStream) {
+	return func(stream *TerminalStream) {
+		stream.resizeC = resizeC
+	}
+}
+
+// NewTerminalStream creates a stream that manages reading and writing
+// data over the provided [websocket.Conn]
+func NewTerminalStream(ws *websocket.Conn, opts ...func(*TerminalStream)) (*TerminalStream, error) {
+	switch {
+	case ws == nil:
+		return nil, trace.BadParameter("required parameter ws not provided")
+	}
+
+	t := &TerminalStream{
+		ws:      ws,
+		encoder: unicode.UTF8.NewEncoder(),
+		decoder: unicode.UTF8.NewDecoder(),
+	}
+
+	for _, opt := range opts {
+		opt(t)
+	}
+
+	return t, nil
+}
+
+// TerminalStream manages the [websocket.Conn] to the web UI
+// for a terminal session.
+type TerminalStream struct {
+	// encoder is used to encode UTF-8 strings.
+	encoder *encoding.Encoder
+	// decoder is used to decode UTF-8 strings.
+	decoder *encoding.Decoder
+
+	// buffer is a buffer used to store the remaining payload data if it did not
+	// fit into the buffer provided by the callee to Read method
+	buffer []byte
+
+	// once ensures that resizeC is closed at most one time
+	once sync.Once
+	// resizeC a channel to forward resize events so that
+	// they happen out of band and don't block reads
+	resizeC chan<- *session.TerminalParams
+
+	// mu protects writes to ws
+	mu sync.Mutex
+	// ws the connection to the UI
+	ws *websocket.Conn
+}
+
+// writeError displays an error in the terminal window.
+func (t *TerminalStream) writeError(err error) error {
+	// Replace \n with \r\n so the message correctly aligned.
+	r := strings.NewReplacer("\n", "\r\n")
```

Comment:
Is it a concern that this would replace "\r\n" with "\r\r\n"? Also since it's only used once why not replace this with strings.ReplaceAll?